time stretching of samples

bgold-cosmos opened the pull request:
This implements time-scaling for sample playback using an overlap-add (OLA) algorithm.

I tweaked the code with speed and flexibility as priorities rather than output sound quality, but for modest scaling amounts (0.8 to 1.3 or so) it usually sounds fairly good.  I'm not aware of any method that will consistently timescale transient-only sounds (most drums) in a "nice" way.

The timescaling _should_ play well with `speed`, `begin`, `end` and similar parameters, but I'm sure I haven't tested every possible combination.

Examples of usage in Tidal would be
```
timescale = pF "timescale"

d1 $ chop 4 $ s "rave" # timescale "[1 4] 2" # cps 0.5

d1 $ s "metal*4" # cps 1
 # up ("c g6 c c6" + "<c ef g c>" - 12)
 # timescale "1 3 1 2"
```

yaxu committed:
Merge pull request #126 from musikinformatik/develop

Sync to master

bgold-cosmos committed:
Merge branch 'master' of https://github.com/musikinformatik/SuperDirt

bgold-cosmos committed:
Merge branch 'develop' of https://github.com/musikinformatik/SuperDirt

telephon committed:
Merge pull request #136 from musikinformatik/develop

Release to master, in sync with tidal v1.1.0

bgold-cosmos committed:
Merge branch 'master' of https://github.com/musikinformatik/SuperDirt

bgold-cosmos committed:
Merge branch 'develop' of https://github.com/musikinformatik/SuperDirt

telephon committed:
Merge pull request #144 from musikinformatik/develop

Small fixes to master (LATEST)

bgold-cosmos committed:
Merge branch 'develop' of https://github.com/musikinformatik/SuperDirt

bgold-cosmos committed:
Merge branch 'master' of https://github.com/musikinformatik/SuperDirt

bgold-cosmos committed:
Merge branch 'develop' of https://github.com/musikinformatik/SuperDirt

bgold-cosmos committed:
Merge branch 'develop' of https://github.com/musikinformatik/SuperDirt

bgold-cosmos committed:
time stretching of samples

initial commit, mostly seems to work but the parameter name(s) and implementation details may change

bgold-cosmos committed:
tweak windowing and fix default stretch

bgold-cosmos committed:
rename to `timescale`, streamline calculations

telephon commented:
just to be sure: have you checked the PitchShift UGen for comparison? Your implementation has the advantage that it could be combined with the https://github.com/musikinformatik/WavesetsEvent Quark, that sounds much better to my ear than overlap add.

In the end, of course we shouldn't load the standard player with extra things, but add a switch when timestretch is set, to save electricity.

bgold-cosmos commented:
Whenever I've tried the PitchShift UGen I've gotten really choppy results regardless of windowSize.  However, I should check again to do a sound comparison, and I'll take a look at the Wavesets stuff.

I wasn't sure how best to conditionally apply the effect.  My understanding is it's not really possible to have true branching in a SynthDef, so would there have to be a whole new set of `dirt_sample_timescale_%_%` SynthDefs that would get called when the event's `~timeScale` is non-nil?  I tried to keep the extra computation to a minimum - all it really is doing is making a pair of phase indices for `BufRd` rather than a single one, and doing a little extra modulation on them so that they're not quite simple LFSaws.  I haven't rigorously compared CPU usage but it didn't seem to add much.

bgold-cosmos committed:
Merge branch 'develop' of https://github.com/musikinformatik/SuperDirt into stretch

bgold-cosmos committed:
Revert "Merge branch 'develop' of https://github.com/musikinformatik/SuperDirt into stretch"

This reverts commit 1d6ce852dcdb144c62058cdfd302a9b93e3eb938, reversing
changes made to 563cda167c166669c448d7db9acf19a199447605.

telephon commented:
Why don't you implement it as an effect?

bgold-cosmos commented:
The current version can compress time too, so to do that as an effect would require looking into the future.  I don't know how to do that except for samples that've already been loaded.  Also, I think working with the sample indexing will be much more efficient, it only ever has two overlapping windows --- whereas as an effect I think it would need to be implemented using a network of staggered delays, some potentially with a very long delay time for large stretches and long samples.  Might be interesting to try both, though...

bgold-cosmos commented:
The current version can compress time too, so to do that as an effect would require looking into the future.  I don't know how to do that except for samples that've already been loaded.  Also, I think working with the sample indexing will be much more efficient, it only ever has two overlapping windows --- whereas as an effect I think it would need to be implemented using a network of staggered delays, some potentially with a very long delay time for large stretches and long samples.  Might be interesting to try both, though...

> On Oct 24, 2019, at 8:37 AM, Julian Rohrhuber <notifications@github.com> wrote:
> 
> Why don't you implement it as an effect?
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <https://github.com/musikinformatik/SuperDirt/pull/149?email_source=notifications&email_token=AB7AIZWKJJLHAJBYJ4575H3QQGQJBA5CNFSM4I7ORIPKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOECFBVOA#issuecomment-545921720>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AB7AIZV5LKKQTUVV6GGUZSTQQGQJBANCNFSM4I7ORIPA>.
>

bgold-cosmos committed:
Rework timescale as separate sample player

The time-scaling effect is now implemented as an alternate sample playback SynthDef that is only called if the `timescale` parameter is used from Tidal.

bgold-cosmos commented:
OK, the timescaling has now been moved to a separate SynthDef so no extra work is done if it isn't used.

telephon commented on synths/core-modules.scd:
```diff
@@ -37,19 +37,36 @@ this may be refacored later.
+						end: ~end,
+						loop: ~loop,
+						pan: ~pan,
+						timescale: ~timescale,
```
we don't need timescale in this case, right?

telephon commented on synths/core-synths.scd:
```diff
@@ -52,6 +50,47 @@ live coding them requires that you have your SuperDirt instance in an environmen
+
+			numFrames = BufFrames.ir(bufnum);
+			windowSize = numFrames * timescale / 29.0;
+			windowSize = windowSize.clip(1000.0*timescale, 4410.0);
```
why not `1000 * timescale, 4410` ? Mixing floats and ints is perfectly fine. Unless you are using specific operators like `rand`, `div`, where it makes a (reasonable) difference.

telephon commented on synths/core-synths.scd:
```diff
@@ -52,6 +50,47 @@ live coding them requires that you have your SuperDirt instance in an environmen
+			phase = (speed.sign * LFSaw.ar(sawrate, 1)).range(begin,end) * numFrames * timescale;
+			index = (phase.div(timescaleStep) - [1.0, 0.0]) * (timescaleStep / timescale - timescaleStep) + phase;
+			windowIndex = phase - (phase.div(timescaleStep) - [1.0, 0.0] * timescaleStep);
+			window = exp(-0.5 * (windowIndex/windowSize - 0.5/0.1)**2);
```
this can be written as:
`exp((5 * windowIndex/windowSize - 2.5) ** 2)`

or maybe better as:
`exp(squared((5 * windowIndex/windowSize - 2.5)))`

why these numbers and not others?

bgold-cosmos replied:
It's a gaussian centered at the middle of the window with a width (sigma) of 0.1 the window size.  The 0.1, though, is really just trial and error to find what sounds good for a wide variety of sounds.  But can definitely combine this into fewer operations.

telephon replied:
maybe it would make sense as a parameter that you can pass, with a good default (in any case it makes sense to avoid unnecessary operations on a signal).

telephon commented on hacks/looper.scd:
```diff
@@ -28,8 +28,9 @@ d1 $ silence
-once $ sound "fooloopReset"
+-- reset buffers. But do only one time (d1 $ silence),
+-- since d1 $ once $ sound "fooloopReset" does not work at my installation (1.4.3)
+d1 $ sound "fooloopReset"
```
Maybe `my installation` is ambiguous here?

bgold-cosmos replied:
I accidentally merged in something from the `develop` branch here, and I thought I undid it but apparently not.  I didn't make this change.

telephon commented on classes/DirtEvent.sc:
```diff
@@ -102,7 +102,7 @@ DirtEvent {
 				~delta * ~legato.value
 			} {
-				unitDuration = unitDuration ? ~delta;
+				unitDuration = (unitDuration ? ~delta) * (~timescale ? 1);
```
it would be clearer and more efficient to write:
```
unitDuration = (unitDuration ? ~delta);
if(~timescale.notNil) { unitDuration = unitDuration * ~timescale };
```

telephon reviewed (changes requested):
The extra complications introduced are worth it, I think! Thank you.

bgold-cosmos committed:
fixes as requested

telephon commented on synths/core-synths.scd:
```diff
@@ -68,13 +68,13 @@ live coding them requires that you have your SuperDirt instance in an environmen
 			numFrames = BufFrames.ir(bufnum);
 			windowSize = numFrames * timescale / 29.0;
-			windowSize = windowSize.clip(1000.0*timescale, 4410.0);
+			windowSize = windowSize.clip(1000*timescale, 4410);
```
Will that work on all sample rates? If not, replace `4410` by `0.1 * SampleRate.ir`).

telephon commented:
btw. sorry, I had overlooked your changes.

telephon commented on synths/core-modules.scd:
```diff
@@ -37,19 +37,35 @@ this may be refacored later.
-					out: ~out
-				])
+				if(~timescale.notNil) {
+					dirtEvent.sendSynth(~stretchInstrument, [
```
I'd change this to:

```
var instrument, args;
...
args = [ ... ]; // the shorter arg list
if(~timescale.notNil) { 
	instrument = ~stretchInstrument;
	args = args.add(\timescale).add(~timescale)
} { 
	instrument = ~instrument 
};

dirtEvent.sendSynth(instrument, args);
…
```
Then you don't need the same list twice.

telephon commented on synths/core-modules.scd:
```diff
@@ -37,19 +37,35 @@ this may be refacored later.
+						out: ~out
+					])
+				} {
+					// argumets could be omitted using getMsgFunc, but for making it easier to understand, we write them out
```
typo: argumets –> arguments

yaxu commented:
Shall we make these smallish changes and @bgold-cosmos ?

bgold-cosmos commented:
I'll refresh my memory on this and look into getting it updated.  A lot of things got put on pause during 2020.  I suspect given all the SuperDirt changes lately it'll be much cleaner to just redo this as a new PR - the semester is just starting Monday here but if I don't do something within a week or so don't feel bad about pinging me again.

telephon commented:
I think it won't be hard to resolve any conflicts.

bgold-cosmos commented:
I've redone this, see #229
